database-commons/src/main/java/io/cdap/plugin/util/RetryUtils.java

```diff
@@ -0,0 +1,309 @@
+        // wrap exception to ensure SQLException-child instances not exposed to contexts without jdbc
+        // driver in classpath
+        String errorMessage =
+          String.format("SQL Exception occurred: [Message='%s', SQLState='%s', ErrorCode='%s'].", e.getMessage(),
+            e.getSQLState(), e.getErrorCode());
+        String sqlState = e.getSQLState();
+        int errorCode = e.getErrorCode();
+        String errorMessageWithDetails = String.format("Error occurred while trying to" +
+          " get schema from database." + "Error message: '%s'. Error code: '%s'. SQLState: '%s'", e.getMessage(),
+            errorCode, sqlState);
+
+        if (!Strings.isNullOrEmpty(externalDocumentationLink)) {
+            if (!errorMessage.endsWith(".")) {
+                errorMessage = errorMessage + ".";
+            }
+            errorMessage = String.format("%s For more details, see %s", errorMessage, externalDocumentationLink);
+        }
+        return ErrorUtils.getProgramFailureException(Strings.isNullOrEmpty(sqlState) ?
+          new ErrorCategory(ErrorCategory.ErrorCategoryEnum.PLUGIN) : getErrorCategoryFromSqlState(sqlState),
+            errorMessage, errorMessageWithDetails, getErrorTypeFromErrorCodeAndSqlState(errorCode, sqlState), true,
+              ErrorCodeType.SQLSTATE, e.getSQLState(), externalDocumentationLink, e);
+    }
+
+    /**
+     * Get the {@link ErrorCategory} for the given SQL state.
+     * Implements generic error categories based on the SQL state.
+     * See <a href="https://en.wikipedia.org/wiki/SQLSTATE">SQLSTATE</a> for more information.
+     * Override this method to provide custom error categories based on the SQL state.
+     *
+     * @param sqlState The SQL state.
+     * @return The {@link ErrorCategory} for the given SQL state.
+     */
+    private static ErrorCategory getErrorCategoryFromSqlState(String sqlState) {
+        if (!Strings.isNullOrEmpty(sqlState) && sqlState.length() >= 2 &&
+          ERROR_CODE_TO_ERROR_CATEGORY.containsKey(sqlState.substring(0, 2))) {
+            return ERROR_CODE_TO_ERROR_CATEGORY.get(sqlState.substring(0, 2));
+        }
+        return new ErrorCategory(ErrorCategory.ErrorCategoryEnum.PLUGIN);
+    }
+
+    /**
+     * Get the {@link ErrorType} for the given error code and SQL state.
+     * Override this method to provide custom error types based on the error code and SQL state.
+     *
+     * @param errorCode The error code.
+     * @param sqlState  The SQL state.
+     * @return The {@link ErrorType} for the given error code and SQL state.
+     */
+    private static ErrorType getErrorTypeFromErrorCodeAndSqlState(int errorCode, String sqlState) {
+        if (!Strings.isNullOrEmpty(sqlState) && sqlState.length() >= 2 &&
+          ERROR_CODE_TO_ERROR_TYPE.containsKey(sqlState.substring(0, 2))) {
+            return ERROR_CODE_TO_ERROR_TYPE.get(sqlState.substring(0, 2));
+        }
+        return ErrorType.UNKNOWN;
+    }
```

Comment:
As I said before this is not needed, we could just create an object of `DBErrorDetailsProvider` and use `DBErrorDetailsProvider#getProgramFailureException` to do all of this stuff.